8349721: Add aarch64 intrinsics for ML-KEM

ferakocz opened the pull request:
By using the aarch64 vector registers the speed of the computation of the ML-KEM algorithms (key generation, encapsulation, decapsulation) can be approximately doubled.

mcpowers commented:
ML-KEM benchmark results of this PR:

```
MLKEM.decapsulate  512 11.80 us/op
MLKEM.decapsulate  768 18.19 us/op 
MLKEM.decapsulate 1024 29.57 us/op
MLKEM.encapsulate  512  8.80 us/op 
MLKEM.encapsulate  768 13.49 us/op  
MLKEM.encapsulate 1024 22.53 us/op
MLKEM.keygen       512  7.49 us/op 
MLKEM.keygen       768 11.22 us/op  
MLKEM.keygen      1024 19.08 us/op 
```

ML-KEM no intrinsics

```
MLKEM.decapsulate  512 31.23 us/op  
MLKEM.decapsulate  768 50.09 us/op 
MLKEM.decapsulate 1024 75.92 us/op
MLKEM.encapsulate  512 22.72 us/op  
MLKEM.encapsulate  768 37.27 us/op 
MLKEM.encapsulate 1024 59.69 us/op
MLKEM.keygen       512 17.95 us/op  
MLKEM.keygen       768 30.95 us/op 
MLKEM.keygen      1024 49.04 us/op
```

adinn commented on src/hotspot/share/opto/library_call.cpp:
```diff
@@ -7600,6 +7614,245 @@ bool LibraryCallKit::inline_chacha20Block() {
+  address stubAddr;
+  const char *stubName;
+  assert(UseKyberIntrinsics, "need Kyber intrinsics support");
+  assert(callee()->signature()->size() == 3, "kyber12To16 has 3 parameters");
```
With this value reset to 4 the ML_DSA test passes for ML_KEM on a debug build.

adinn reviewed (commented):
@ferakocz Thanks for another very good piece of work which appears to me to be functioning correctly and performantly.

The PR suffers from the same problems as the original ML_DSA one i.e. The mapping of data to registers and the overall structure of the generated code and its relation to the related Java code/the original algorithms will be hard for a maintainer to identify. I have reworked your patch to use vector sequences in this [draft PR](https://github.com/openjdk/jdk/pull/24419) in very much the same way as was done for the ML_DSA PR. This has significantly abstracted and clarified the register mappings that are in use in each kyber generator and has also made the higher level structure of the generated code much easier to follow.

Note that my rework of the generation routines was applied to your original PR after rebasing it on master. Before updating the kyber routines I also generalized a few of the VSeq<N> methods that benefit from being shared by both kyber and dilithium, most notably the montmul routines, and I added a few extra helpers.

The reworked version passes the ML_KEM functional test and gives similar performance improvements for the ML_KEM micro benchmark. The generated code does differ in a few places from what your original patch generates but only superficially - most notable is that a few loads/stores that rely on continued post-increments in the original instead use a constant offset or an add/load pair in the reworked code. This makes a very minor difference to code size and does not seem to affect performance.

I would like you to rework your PR to incorporate these changes because I believe it will make a big difference to maintainability.  n.b. it may be easier to integrate my changes by diffing your branch and mine and applying the resulting change set rather than trying to merge the changes. Please let me know if you have problems with the integration and need help.

I still have some further review comments and would also like to see more commenting to explain what the code is doing. However, I think it will be easier to do that after this rework has been integrated into your PR.

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4840,40 +4952,107 @@ class StubGenerator: public StubCodeGenerator {
+    assert(vs_disjoint(va, vtmp), "va and vtmp overlap");
+    assert(!va.is_constant(), "output vector must identify 2 different registers");
+
+    // schedule 2 streams of i<nstructions across the vector sequences
```
typo: i<nstructions -> instructions

adinn replied:
```suggestion
    // schedule 2 streams of instructions across the vector sequences
```

adinn replied:
```suggestion
    // schedule 2 streams of instructions across the vector sequences
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    // On each level, we fill up the vector registers in such a way that the
+    // array elements that need to be multiplied by the zetas be in one
+    // set of vector registers while the corresponding ones that don't need to
+    // be multiplied, in another set. We can do 32 Montgomery multiplications
```
grammar: be multiplied, in another set. --> be multiplied are in another set.

adinn replied:
```suggestion
    // be multiplied are in another set. We can do 32 Montgomery multiplications
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+
+    // level 4
+    vs_ldpq(vq, kyberConsts);
+    int offsets3[8] = { 0, 32, 64, 96, 128, 160, 192, 224 };
```
I'd like to add comment here to explain the coefficient grouping
and likewise at level 5 and 6. So here we have:
// Up to level 3 the coefficients multiplied by or added/subtracted
// to the zetas occur in discrete blocks whose size is some multiple
// of 32. At level 4 coefficients occur in 8 discrete blocks of size 16
// so they are loaded using employing an ldr at 8 distinct offsets.

adinn replied:
I'd like to add comment here to explain the coefficient grouping
and likewise at level 5 and 6. So here we have:
```suggestion
    // Up to level 3 the coefficients multiplied by or added/subtracted
    // to the zetas occur in discrete blocks whose size is some multiple
    // of 32. At level 4 coefficients occur in 8 discrete blocks of size 16
    // so they are loaded using an ldr at 8 distinct offsets.
    int offsets3[8] = { 0, 32, 64, 96, 128, 160, 192, 224 };
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+
+    // level 0
+    vs_ldpq(vq, kyberConsts);
+    int offsets4[4] = { 0, 32, 64, 96 };
```
Again a comment
// At level 0 related coefficients occur in discrete blocks of size 4 so
// need to be loaded interleaved using an ld2 operation with arrangement 4S

adinn replied:
Again a comment
```suggestion
     // At level 6 related coefficients occur in discrete blocks of size 4 so
     // need to be loaded interleaved using an ld2 operation with arrangement 4S
     int offsets4[4] = { 0, 32, 64, 96 };
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+                            vs_front(vs2), vs_back(vs2), vtmp, vq);
+    vs_st2_indexed(vs1, __ T4S, coeffs, tmpAddr, 384, offsets4);
+
+    // level 1
```
Again a comment
// At level 1 related coefficients occur in discrete blocks of size 8 so
// need to be loaded interleaved using an ld2 operation with arrangement 2D

adinn replied:
Again a comment
```suggestion
    // level 1
    // At level 1 related coefficients occur in discrete blocks of size 8 so
    // need to be loaded interleaved using an ld2 operation with arrangement 2D
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    vs_st2_indexed(vs1, __ T2D, coeffs, tmpAddr, 384, offsets4);
+
+    // level 2
+    int offsets3[8] = { 0, 32, 64, 96, 128, 160, 192, 224 };
```
Again
// At level 3 coefficients occur in 8 discrete blocks of size 16
// so they are loaded using employing an ldr at 8 distinct offsets.

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    kyber_montmul64(vs2, vs1, vq3, vtmp, vq);
+    __ add(tmpAddr, coeffs, 0);
+    store64shorts(vs2, tmpAddr);
+
```
I'd like to make explicit the fact that we have avoided doing an add here (and in the next two cases) by adding a commented out generation step i.e. at this line insert

    // __ add(tmpAddr, coeffs, 128); // unneeded as implied by preceding load

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    kyber_montmul64(vs2, vs1, vq3, vtmp, vq);
+    __ add(tmpAddr, coeffs, 128);
+    store64shorts(vs2, tmpAddr);
+
```
Likewise add:
    // __ add(tmpAddr, coeffs, 256); // unneeded as implied by preceding load

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    __ add(tmpAddr, coeffs, 256);
+    store64shorts(vs2, tmpAddr);
+
+    load64shorts(vs1, tmpAddr);
```
Likewise add:

    // __ add(tmpAddr, coeffs, 384); // unneeded as implied by preceding load

adinn replied:
Likewise insert:
```suggestion
    // __ add(tmpAddr, coeffs, 384); // unneeded as implied by preceding store
    load64shorts(vs1, tmpAddr);
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    // load 96 (6 x 16B) byte values
+    vs_ld3_post(vin, __ T16B, condensed);
+
+    // expand groups of input bytes in vin to shorts in va and vb
```
I's like to expand on the data layouts here so that maintenance engineers don't have to work it out every time they look at it. So, I would like to replace this comment as follows

    // The front half of sequence vin (vin[0], vin[1] and vin[2])
    // holds 48 (16x3) contiguous bytes from memory striped
    // horizontally across each of the 16 byte lanes. Equivalently,
    // that is 16 pairs of 12-bit integers. Likewise the back half
    // holds the next 48 bytes in the same arrangement.

    // Each vector in the front half can also be viewed as a vertical
    // strip across the 16 pairs of 12 bit integers. Each byte in
    // vin[0] stores the low 8 bits of the first int in a pair. Each
    // byte in vin[1] stores the high 4 bits of the first int and the
    // low 4 bits of the second int. Each byte in vin[2] stores the
    // high 8 bits of the second int. Likewise the vectors in second
    // half.

    // Converting the data to 16-bit shorts requires first of all
    // expanding each of the 6 x 16B vectors into 6 corresponding
    // pairs of 8H vectors. Mask, shift and add operations on the
    // resulting vector pairs can be used to combine 4 and 8 bit
    // parts of related 8H vector elements.
    //
    // The middle vectors (vin[2] and vin[5]) are actually expanded
    // twice, one copy manipulated to provide the lower 4 bits
    // belonging to the first short in a pair and another copy
    // manipulated to provide the higher 4 bits belonging to the
    // second short in a pair. This is why the the vector sequences va
    // and vb used to hold the expanded 8H elements are of length 8.

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    __ ushll2(va[3], __ T8H, vin[1], __ T16B, 0);
+    __ ushll(va[4], __ T8H, vin[1], __ T8B, 0);
+    __ ushll2(va[5], __ T8H, vin[1], __ T16B, 0);
+
```
Insert here
    // Likewise expand vin[3] into vb[0:1], and vin[4] into vb[2:3]
    // and vb[4:5]

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    __ shl(vb[2], __ T8H, vb[2], 8);
+    __ shl(vb[3], __ T8H, vb[3], 8);
+
+    // expand remaining input bytes in vin to shorts in va and vb
```
To make this clearer it should say

    // Expand vin[2] into va[6:7] and vin[5] into vb[6:7] but this
    // time pre-shifted by 4 to ensure top bits of input 12-bit int
    // are in bit positions [4..11].

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    __ ushll(vb[6], __ T8H, vin[5], __ T8B, 4);
+    __ ushll2(vb[7], __ T8H, vin[5], __ T16B, 4);
+
+    // split the duplicated 8 bit values into two distinct 4 bit
```
To make this clearer it should say

    // mask hi 4 bits of the 1st 12-bit int in a pair from copy1 and
    // shift lo 4 bits of the 2nd 12-bit int in a pair to the bottom of
    // copy2

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    __ ushr(vb[4], __ T8H, vb[4], 4);
+    __ ushr(vb[5], __ T8H, vb[5], 4);
+
+    // sum resulting short values into the front halves of va and
```
This should be replaced to clarify details of the ordering for summing and grouping

    // sum hi 4 bits and lo 8 bits of the 1st 12-bit int in each pair and
    // hi 8 bits plus lo 4 bits of the 2nd 12-bit int in each pair

    // n.b. the ordering ensures: i) inputs are consumed before they
    // are overwritten ii) the order of 16-bit results across successive
    // pairs of vectors in va and then vb reflects the order of the
    // corresponding 12-bit inputs

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    vs_ld3_post(vs_front(vin), __ T16B, condensed);
+    vs_ld3(vs_back(vin), __ T8B, condensed);
+
+    // expand groups of input bytes in vin to shorts in va and vb
```
Clarify as follows

    // if anything is left it should be a final 72 bytes of input
    // i.e. a final 48 12-bit values. so we handle this by loading
    // load 48 bytes into all 16B lanes of front(vin) and only 24
    // bytes into the lower 8B lane of back(vin)

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    __ ushll(va[4], __ T8H, vin[1], __ T8B, 0);
+    __ ushll2(va[5], __ T8H, vin[1], __ T16B, 0);
+
+    __ ushll(vb[0], __ T8H, vin[3], __ T8B, 0);
```
Add a comment

    // This time expand just the lower 8 lanes of vin[3] into vb[0], and
    // just the lower 8 lanes of vin[4] into vb[2] and vb[4]

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    __ ushll(vb[2], __ T8H, vin[4], __ T8B, 0);
+    __ ushll(vb[4], __ T8H, vin[4], __ T8B, 0);
+
+    // offset duplicated elements in va and vb by 8
```
As before clarify as follows

    // shift lo byte of copy 1 of the middle stripe into the high byte

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    __ shl(va[3], __ T8H, va[3], 8);
+    __ shl(vb[2], __ T8H, vb[2], 8);
+
+    // expand remaining input bytes in vin to shorts in va and vb
```
Again improve this comment

    // expand vin[2] into va[6:7] and lower 8 lanes of vin[5] into
    // vb[6] pre-shifted by 4 to ensure top bits of the input 12-bit
    // int are in bit positions [4..11].

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    __ ushll2(va[7], __ T8H, vin[2], __ T16B, 4);
+    __ ushll(vb[6], __ T8H, vin[5], __ T8B, 4);
+
+    // split the duplicated 8 bit values into two distinct 4 bit
```
Once again update

    // mask hi 4 bits of each 1st 12-bit int in pair from copy1 and
    // shift lo 4 bits of each 2nd 12-bit int in pair to bottom of
    // copy2

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    __ andr(vb[2], __ T16B, vb[2], v31);
+    __ ushr(vb[4], __ T8H, vb[4], 4);
+
+    // sum resulting short values into the front halves of va and
```
Again update to provide more detail

    // sum hi 4 bits and lo 8 bits of each 1st 12-bit int in pair and
    // hi 8 bits plus lo 4 bits of each 2nd 12-bit int in pair

    // n.b. ordering ensures: i) inputs are consumed before they are
    // overwritten ii) order of 16-bit results across succsessive
    // pairs of vectors in va and then lower half of vb reflects order
    // of corresponding 12-bit inputs

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    __ addv(vb[0], __ T8H, vb[0], vb[2]);
+    __ addv(vb[1], __ T8H, vb[4], vb[6]);
+
+    // store results interleaved as shorts
```
Change to

    // store 48 results interleaved as shorts

adinn commented:
@ferakocz Hi Ferenc. Thank you for adjusting the code as requested and even more so for the extra clean-ups you added which I very much appreciate.

I have added suggestions for some extra/modified commenting to clarify certain details of what is being generated that were not 100% clear to me when I first read/restructured the code. They may seem a bit obvious but I want to ensure that any maintainer who needs to review the code can assimilate it quickly (including me if/when I revisit it in 12 months time).

Mostly my recommendations for upgrading of comments is complete and I believe little more will be needed to sign off this PR. However, I still want to check through a few parts of the code that I have not fully cross-checked against the Java routines (e.g. the Barrett reductions). I'll try to do that asap but it will probably be a few days from now.

Thanks again for your help in improving this code.

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    VSeq<2> vs2_2(26);
+    FloatRegister vs2_3 = v29;
+
+    // we also need a pair of corresponding constant sequences
+    VSeq<8> vc1_1(30, 0);
+    VSeq<2> vc1_2(30, 0);
+
+    FloatRegister vc1_3 = v30;
+    VSeq<8> vc2_1(31, 0);
+    VSeq<2> vc2_2(31, 0);
+    FloatRegister vc2_3 = v31;
```
```suggestion
    // we also need a pair of corresponding constant sequences
    VSeq<8> vc1_1(30, 0);       // kyber_q
    VSeq<2> vc1_2(30, 0);
    FloatRegister vc1_3 = v30;

    VSeq<8> vc2_1(31, 0);       // kyberBarrettMultiplier
    VSeq<2> vc2_2(31, 0);
    FloatRegister vc2_3 = v31;
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+
+    // load q and the multiplier for the Barrett reduction
+    __ add(kyberConsts, kyberConsts, 16);
+    __ ldpq(vc1_3, vc2_3, kyberConsts);
```
```suggestion
    __ ldpq(vc1_3, vc2_3, kyberConsts); // kyber_q, kyberBarrettMultiplier
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+      if (i < 2) {
+        __ ldr(vs1_3, __ Q, __ post(coeffs, 16));
+      }
+      vs_sqdmulh(vs2_1, __ T8H, vs1_1, vc2_1);
```
```suggestion
      vs_sqdmulh(vs2_1, __ T8H, vs1_1, vc2_1); // vs2 <- (2 * vs1 * kyberBarrettMultiplier) >> 16
```

adinn replied:
```suggestion
      vs_sqdmulh(vs2_1, __ T8H, vs1_1, vc2_1); // vs2 <- (2 * vs1 * kyberBarrettMultiplier) >> 16
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+      if (i < 2) {
+        __ sqdmulh(vs2_3, __ T8H, vs1_3, vc2_3);
+      }
+      vs_sshr(vs2_1, __ T8H, vs2_1, 11);
```
```suggestion
      vs_sshr(vs2_1, __ T8H, vs2_1, 11); // vs2 <- (vs1 * kyberBarrettMultiplier) >> 26
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+      if (i < 2) {
+        __ sshr(vs2_3, __ T8H, vs2_3, 11);
+      }
+      vs_mlsv(vs1_1, __ T8H, vs2_1, vc1_1);
```
```suggestion
      vs_mlsv(vs1_1, __ T8H, vs2_1, vc1_1); // vs1 <- vs1 - vs2 * kyber_q
```

adinn commented on src/hotspot/cpu/aarch64/register_aarch64.hpp:
```diff
@@ -457,27 +458,89 @@ template<int N> class VSeq {
+  return false;
+}
+
+// convenience methods for splitting 8-way of 4-way vector register
```
```suggestion
// convenience methods for splitting 8-way or 4-way vector register
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    vs_str_indexed(vs3, __ Q, coeffs, 256 + 16, offsets3);
+
+    // level 5
+    vs_ldpq(vq, kyberConsts);
```
```suggestion
    vs_ldpq(vq, kyberConsts);
    // At level 5 related coefficients occur in discrete blocks of size 8 so
    // need to be loaded interleaved using an ld2 operation with arrangement 2D
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    kyber_montmul32_sub_add(vs_even(vs1), vs_odd(vs1), vs_front(vs2), vtmp, vq);
+    vs_st2_indexed(vs1, __ T2D, coeffs, tmpAddr, 384, offsets4);
+
+    // level 6
```
```suggestion
    // level 6
    // At level 6 related coefficients occur in discrete blocks of size 4 so
    // need to be loaded interleaved using an ld2 operation with arrangement 4S
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+                            vs_front(vs2), vs_back(vs2), vtmp, vq);
+    vs_st2_indexed(vs1, __ T2D, coeffs, tmpAddr, 384, offsets4);
+
+    // level 2
```
Again
```suggestion
    // level 2
    // At level 2 coefficients occur in 8 discrete blocks of size 16
    // so they are loaded using employing an ldr at 8 distinct offsets.
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    vs_mlsv(vs1, __ T8H, vs2, vq1);
+    vs_str_indexed(vs1, __ Q, coeffs, 256, offsets3);
+
+    // level 3
```
```suggestion
    // level 3
    // At level 1 related coefficients occur in discrete blocks of size 8 so
    // need to be loaded interleaved using an ld2 operation with arrangement 2D
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    __ add(tmpAddr, coeffs, 0);
+    store64shorts(vs2, tmpAddr);
+
+    load64shorts(vs1, tmpAddr);
```
I'd like to make explicit the fact that we have avoided doing an add here (and in the next two cases) by adding a commented out generation step i.e. at this line insert
```suggestion
    // __ add(tmpAddr, coeffs, 128); // unneeded as implied by preceding load
    load64shorts(vs1, tmpAddr);
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    __ add(tmpAddr, coeffs, 128);
+    store64shorts(vs2, tmpAddr);
+
+    load64shorts(vs1, tmpAddr);
```
Likewise insert:
```suggestion
    // __ add(tmpAddr, coeffs, 256); // unneeded as implied by preceding load
    load64shorts(vs1, tmpAddr);
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+
+    VSeq<4> vs1(0), vs2(4);  // 4 sets of 8x8H inputs/outputs/tmps
+    VSeq<4> vs3(16), vs4(20);
+    VSeq<2> vq(30);          // pair of constants for montmul
```
```suggestion
    VSeq<2> vq(30);          // pair of constants for montmul: q, qinv
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    VSeq<4> vs3(16), vs4(20);
+    VSeq<2> vq(30);          // pair of constants for montmul
+    VSeq<2> vz(28);          // pair of zetas
+    VSeq<4> vc(27, 0);       // constant sequence for montmul
```
```suggestion
    VSeq<4> vc(27, 0);       // constant for montmul: montRSquareModQ
```

ferakocz commented:
@adinn Hi, Andrew,
 I think I addressed all of your comment improvement comments, in most cases I just changed them as you suggested. Thanks a lot for the thorough review!

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    __ BIND(kyberNttMult_loop);
+    // load 16 zetas
+    vs_ldpq_post(vz, zetas);
+    // load 2 sets of 32 coefficients from the two input arrays
```
```suggestion
    // load 2 sets of 32 coefficients from the two input arrays
    // interleaved as shorts. i.e. pairs of shorts adjacent in memory
    // are striped across pairs of vector registers
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    // load 16 zetas
+    vs_ldpq_post(vz, zetas);
+    // load 2 sets of 32 coefficients from the two input arrays
+    vs_ld2_post(vs_front(vs1), __ T8H, ntta);
+    vs_ld2_post(vs_back(vs1), __ T8H, nttb);
+    vs_ld2_post(vs_front(vs4), __ T8H, ntta);
+    vs_ld2_post(vs_back(vs4), __ T8H, nttb);
```
```suggestion
    vs_ld2_post(vs_front(vs1), __ T8H, ntta); // <a0, a1> x 8H
    vs_ld2_post(vs_back(vs1), __ T8H, nttb);  // <b0, b1> x 8H
    vs_ld2_post(vs_front(vs4), __ T8H, ntta); // <a2, a3> x 8H
    vs_ld2_post(vs_back(vs4), __ T8H, nttb);  // <b2, b3> x 8H
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    vs_ld2_post(vs_back(vs1), __ T8H, nttb);
+    vs_ld2_post(vs_front(vs4), __ T8H, ntta);
+    vs_ld2_post(vs_back(vs4), __ T8H, nttb);
+    // montmul the first and second pair of values loaded into vs1
+    // in order and then with one pair reversed storing the  two
+    // results in vs3
```
```suggestion
    // compute 4 montmul cross-products for pairs (a0,a1) and (b0,b1)
    // i.e. montmul the first and second halves of vs1 in order and
    // then with one sequence reversed storing the two results in vs3
    //
    // vs3[0] <- montmul(a0, b0)
    // vs3[1] <- montmul(a1, b1)
    // vs3[2] <- montmul(a0, b1)
    // vs3[3] <- montmul(a1, b0)
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    kyber_montmul16(vs_front(vs3), vs_front(vs1), vs_back(vs1), vs_front(vs2), vq);
+    kyber_montmul16(vs_back(vs3),
+                    vs_front(vs1), vs_reverse(vs_back(vs1)), vs_back(vs2), vq);
+    // montmul the first and second pair of values loaded into vs4
+    // in order and then with one pair reversed storing the two
+    // results in vs1
```
```suggestion
    // compute 4 montmul cross-products for pairs (a2,a3) and (b2,b3)
    // i.e. montmul the first and second halves of vs4 in order and
    // then with one sequence reversed storing the two results in vs1
    //
    // vs1[0] <- montmul(a2, b2)
    // vs1[1] <- montmul(a3, b3)
    // vs1[2] <- montmul(a2, b3)
    // vs1[3] <- montmul(a3, b2)
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    kyber_montmul16(vs_front(vs1), vs_front(vs4), vs_back(vs4), vs_front(vs2), vq);
+    kyber_montmul16(vs_back(vs1),
+                    vs_front(vs4), vs_reverse(vs_back(vs4)), vs_back(vs2), vq);
+    // for each pair of results pick the second value in the first
+    // pair to create a sequence that we montmul by the zetas
+    // i.e. we want sequence <vs3[1], vs1[1]>
```
```suggestion
    // montmul result 2 of each cross-product i.e. (a1*b1, a3*b3) by a zeta.
    // We can schedule two montmuls at a time if we use a suitable vector
    // sequence <vs3[1], vs1[1]>.
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    // i.e. we want sequence <vs3[1], vs1[1]>
+    int delta = vs1[1]->encoding() - vs3[1]->encoding();
+    VSeq<2> vs5(vs3[1], delta);
+    kyber_montmul16(vs5, vz, vs5, vs_front(vs2), vq);
```
```suggestion
    // vs3[1] <- montmul(montmul(a1, b1), z0)
    // vs1[1] <- montmul(montmul(a3, b3), z1)
    kyber_montmul16(vs5, vz, vs5, vs_front(vs2), vq);
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    int delta = vs1[1]->encoding() - vs3[1]->encoding();
+    VSeq<2> vs5(vs3[1], delta);
+    kyber_montmul16(vs5, vz, vs5, vs_front(vs2), vq);
+    // add results in pairs storing in vs3
```
```suggestion
    // add results in pairs storing in vs3
    // vs3[0] <- montmul(a0, b0) + montmul(montmul(a1, b1), z0);
    // vs3[1] <- montmul(a0, b1) + montmul(a1, b0);
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    kyber_montmul16(vs5, vz, vs5, vs_front(vs2), vq);
+    // add results in pairs storing in vs3
+    vs_addv(vs_front(vs3), __ T8H, vs_even(vs3), vs_odd(vs3));
+    vs_addv(vs_back(vs3), __ T8H, vs_even(vs1), vs_odd(vs1));
```
```suggestion
    // vs3[2] <- montmul(a2, b2) + montmul(montmul(a3, b3), z1);
    // vs3[3] <- montmul(a2, b3) + montmul(a3, b2);
    vs_addv(vs_back(vs3), __ T8H, vs_even(vs1), vs_odd(vs1));
```

adinn commented on src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:
```diff
@@ -4901,14 +5080,1120 @@ class StubGenerator: public StubCodeGenerator {
+    vs_addv(vs_back(vs3), __ T8H, vs_even(vs1), vs_odd(vs1));
+    // montmul result by constant vc and store result in vs1
+    kyber_montmul32(vs1, vs3, vc, vs2, vq);
+    // store the four results as two interleaved pairs of
+    // quadwords
```
```suggestion
    // store back the two pairs of result vectors de-interleaved as 8H elements
    // i.e. storing each pairs of shorts striped across a register pair adjacent
    // in memory
```

adinn commented:
@ferakocz

Hi Ferenc,

Sorry, but I still had a few comments to add to the KyberNTTMult routine to clarify exactly how the load, compute and store operations relate to the original Java source. That's the only remaining code that I felt needed further clarification for maintainers. So, after you work through them I can approve the PR.

ferakocz commented:
> @ferakocz
> 
> Hi Ferenc,
> 
> Sorry, but I still had a few comments to add to the KyberNTTMult routine to clarify exactly how the load, compute and store operations relate to the original Java source. That's the only remaining code that I felt needed further clarification for maintainers. So, after you work through them I can approve the PR.

No problem , it was easy to make the changes. Thanks again!

adinn commented:
@ferakocz I reran test jtreg:test/jdk/sun/security/provider/acvp/Launcher.java and hit a Java assertion:

```
>> ML-KEM-512 encapsulation
1 STDERR:
java.lang.AssertionError
	at java.base/com.sun.crypto.provider.ML_KEM.twelve2Sixteen(ML_KEM.java:1371)
	at java.base/com.sun.crypto.provider.ML_KEM.decodePoly(ML_KEM.java:1408)
	at java.base/com.sun.crypto.provider.ML_KEM.decodeVector(ML_KEM.java:1337)
	at java.base/com.sun.crypto.provider.ML_KEM.kPkeEncrypt(ML_KEM.java:712)
	at java.base/com.sun.crypto.provider.ML_KEM.encapsulate(ML_KEM.java:555)
	at java.base/com.sun.crypto.provider.ML_KEM_Impls$K.implEncapsulate(ML_KEM_Impls.java:134)
	at java.base/sun.security.provider.NamedKEM$KeyConsumerImpl.engineEncapsulate(NamedKEM.java:124)
	at java.base/javax.crypto.KEM$Encapsulator.encapsulate(KEM.java:265)
	at java.base/javax.crypto.KEM$Encapsulator.encapsulate(KEM.java:225)
	at ML_KEM_Test.encapDecapTest(ML_KEM_Test.java:98)
	at ML_KEM_Test.run(ML_KEM_Test.java:41)
	at Launcher.run(Launcher.java:160)
	at Launcher.main(Launcher.java:122)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:565)
	at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:335)
	at java.base/java.lang.Thread.run(Thread.java:1447)

JavaTest Message: Test threw exception: java.lang.AssertionError
JavaTest Message: shutting down test
```

The offending code is this:

    private void twelve2Sixteen(byte[] condensed, int index,
                                short[] parsed, int parsedLength) {
        int i = parsedLength / 64;
        int remainder = parsedLength - i * 64;
        if (remainder != 0) {
            i++;
        }
        assert (((remainder != 0) && (remainder != 48)) || <== assert here
            index + i * 96 > condensed.length);

I believe the logic is reversed here i.e. it should be:

        assert ((remainder == 0) || (remainder == 48)) &&
            index + i * 96 <= condensed.length);

Does that sound right?

ferakocz commented:
> @ferakocz I reran test jtreg:test/jdk/sun/security/provider/acvp/Launcher.java and hit a Java assertion:
> 
> ```
> >> ML-KEM-512 encapsulation
> 1 STDERR:
> java.lang.AssertionError
> 	at java.base/com.sun.crypto.provider.ML_KEM.twelve2Sixteen(ML_KEM.java:1371)
> 	at java.base/com.sun.crypto.provider.ML_KEM.decodePoly(ML_KEM.java:1408)
> 	at java.base/com.sun.crypto.provider.ML_KEM.decodeVector(ML_KEM.java:1337)
> 	at java.base/com.sun.crypto.provider.ML_KEM.kPkeEncrypt(ML_KEM.java:712)
> 	at java.base/com.sun.crypto.provider.ML_KEM.encapsulate(ML_KEM.java:555)
> 	at java.base/com.sun.crypto.provider.ML_KEM_Impls$K.implEncapsulate(ML_KEM_Impls.java:134)
> 	at java.base/sun.security.provider.NamedKEM$KeyConsumerImpl.engineEncapsulate(NamedKEM.java:124)
> 	at java.base/javax.crypto.KEM$Encapsulator.encapsulate(KEM.java:265)
> 	at java.base/javax.crypto.KEM$Encapsulator.encapsulate(KEM.java:225)
> 	at ML_KEM_Test.encapDecapTest(ML_KEM_Test.java:98)
> 	at ML_KEM_Test.run(ML_KEM_Test.java:41)
> 	at Launcher.run(Launcher.java:160)
> 	at Launcher.main(Launcher.java:122)
> 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:565)
> 	at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:335)
> 	at java.base/java.lang.Thread.run(Thread.java:1447)
> 
> JavaTest Message: Test threw exception: java.lang.AssertionError
> JavaTest Message: shutting down test
> ```
> 
> The offending code is this:
> 
> ```
> private void twelve2Sixteen(byte[] condensed, int index,
>                             short[] parsed, int parsedLength) {
>     int i = parsedLength / 64;
>     int remainder = parsedLength - i * 64;
>     if (remainder != 0) {
>         i++;
>     }
>     assert (((remainder != 0) && (remainder != 48)) || <== assert here
>         index + i * 96 > condensed.length);
> ```
> 
> I believe the logic is reversed here i.e. it should be:
> 
> ```
>     assert ((remainder == 0) || (remainder == 48)) &&
>         index + i * 96 <= condensed.length);
> ```
> 
> Does that sound right?

Aarrrrgh, yes. I forgot to negate that condition when I went from throwing an exception to assert, and I also thought, incorrectly, that -ea would enable my assertions when I tested  :-( .
Thanks a lot for catching it!

adinn reviewed (approved):
@ferakocz I reran the test and the perf test and all appears to be good. Nice work!

ferakocz commented:
> @ferakocz I reran the test and the perf test and all appears to be good. Nice work!

@adinn Thanks, Andrew, for all the help, it really looks nicer than it looked before your review! Would you /sponsor the integration?

ferakocz commented:
/integrate

adinn commented:
/sponsor

iwanowww commented on src/hotspot/cpu/aarch64/vm_version_aarch64.cpp:
```diff
@@ -703,6 +714,7 @@ void VM_Version::initialize_cpu_information(void) {
   get_compatible_board(_cpu_desc + desc_len, CPU_DETAILED_DESC_BUF_SIZE - desc_len);
   desc_len = (int)strlen(_cpu_desc);
   snprintf(_cpu_desc + desc_len, CPU_DETAILED_DESC_BUF_SIZE - desc_len, " %s", _features_string);
+  fprintf(stderr, "_features_string = \"%s\"", _features_string);
```
Was this line added by mistake? Looks like a leftover.

ferakocz replied:
@iwanowww  Oooops, yes. Addressed in [https://github.com/openjdk/jdk/pull/24717](url)  together with another forgotten (though not strictly necessary) change. Thanks for catching it!
Could you review that (really short) PR?

adinn replied:
@ferakocz That link has the right label but leads to the wrong place. It should point [here](https://github.com/openjdk/jdk/pull/24717).

adinn replied:
@iwanowww Thanks for the heads up!
